### PR TITLE
feat(handler): implements handle workflow registered event

### PR DESCRIPTION
**Pull request by MStreet3**
Implements the handler for workflow registered events.

### CODE REVIEW

File: core/services/workflows/syncer/handler.go [hunk 1]
```diff
@@ -135,12 +158,119 @@ func (h *eventHandler) Handle(ctx context.Context, event WorkflowRegistryEvent)
...
 // workflowRegisteredEvent handles the WorkflowRegisteredEvent event type.
-// TODO: Implement this method
 func (h *eventHandler) workflowRegisteredEvent(
-	_ context.Context,
-	_ WorkflowRegistryEvent,
+	ctx context.Context,
+	event WorkflowRegistryEvent,
 ) error {
-	return ErrNotImplemented
+	payload, ok := event.Data.(WorkflowRegistryWorkflowRegisteredV1)
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/handler.go]: Nit: move this up into whatever calls the handler and make the handler event type specific; it'll clutter this method less, and you won't need a guard clause in every handler

File: core/services/workflows/syncer/handler.go [hunk 2]
```diff
@@ -135,12 +158,119 @@ func (h *eventHandler) Handle(ctx context.Context, event WorkflowRegistryEvent)
...
+	cma := h.emitter.With(
+		platform.KeyWorkflowID, wfID,
+		platform.KeyWorkflowName, payload.WorkflowName,
+		platform.KeyWorkflowOwner, hex.EncodeToString(payload.WorkflowOwner),
+	)
+
+	// Download the contents of binaryURL, configURL and secretsURL and cache them locally.
+	binary, err := h.fetcher(ctx, payload.BinaryURL)
+	if err != nil {
+		logCustMsg(ctx, cma, fmt.Sprintf("failed to fetch binary: %v", err), h.lggr)
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/handler.go]: Is there a way to put the custom logging up one level? That way handlers can just return an error, and the thing calling the handler handles the error

One hitch that might make this difficult is the fact that your emitter is presumably event specific.

**Review by cedric-cordenier — DISMISSED** [review]

**Review by ettec — DISMISSED** [review]

**Review by ettec — DISMISSED** [review]

**Review by cedric-cordenier — DISMISSED** [review]

File: core/services/workflows/syncer/handler.go [hunk 3]
```diff
@@ -170,32 +311,47 @@ func (h *eventHandler) workflowActivatedEvent(
...
 	if _, err := h.orm.Update(ctx, hash, string(secrets)); err != nil {
-		return err
+		return fmt.Errorf("failed to update secrets: %w", err)
 	}
 
 	return nil
 }
+
+// sha calculates the sha256 hash of the wasm, config and secretsURL to determine the workflow ID.
+func sha(wasm, config, secretsURL []byte) string {
```
Comment (jmank88) [inline core/services/workflows/syncer/handler.go]:
```suggestion
// workflowID returns a hex encoded sha256 hash of the wasm, config and secretsURL.
func workflowID(wasm, config, secretsURL []byte) string {
```

File: core/services/workflows/syncer/handler.go [hunk 4]
```diff
@@ -170,32 +311,47 @@ func (h *eventHandler) workflowActivatedEvent(
...
+	sum.Write(config)
+	sum.Write(secretsURL)
+	return hex.EncodeToString(sum.Sum(nil))
+}
+
+// logCustMsg emits a custom message to the external sink and logs an error if that fails.
+func logCustMsg(ctx context.Context, cma custmsg.MessageEmitter, msg string, log logger.Logger) {
+	err := cma.Emit(ctx, msg)
+	if err != nil {
+		log.Errorf("failed to send custom message with msg: %s, err: %v", msg, err)
```
Comment (jmank88) [inline core/services/workflows/syncer/handler.go]: Will it be problematic that this log line will point here rather than the caller?

Comment (MStreet3) [reply core/services/workflows/syncer/handler.go]: hm, potentially, the context will be in the message, but do you suggest I just return the error instead and let the caller handle it?

Comment (jmank88) [reply core/services/workflows/syncer/handler.go]: If handling in the caller is annoying, then you can just do:
```suggestion
		log.Helper(1).Errorf("failed to send custom message with msg: %s, err: %v", msg, err)
```
But it constructs a new logger, so not great if it's on the hot path. nbd if this is truly exceptional though.

**Review by jmank88 — APPROVED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Review by justinkaseman — APPROVED** [review]

**Pull request merged**